language/evm/move-to-yul/src/solidity_ty.rs

```diff
@@ -0,0 +1,627 @@
+                    generate_tuple(&tys) // TODO: translate into tuple type?
+                }
+            }
+            TypeParameter(_)
+            | Reference(_, _)
+            | Fun(_, _)
+            | TypeDomain(_)
+            | ResourceDomain(_, _, _)
+            | Error
+            | Var(_) => {
+                panic!("unexpected field type")
+            }
+        }
+    }
+
+    /// Create a default solidity signature from a move function signature
+    pub fn create_default_solidity_signature(&self, fun: &FunctionEnv<'_>) -> SoliditySignature {
+        let fun_name = fun.symbol_pool().string(fun.get_name()).to_string();
+        let mut para_type_lst = vec![];
+        for move_ty in fun.get_parameter_types() {
+            let solidity_ty = self.translate_move_to_solidity_type(&move_ty); // implicit mapping from a move type to a solidity type
+            para_type_lst.push((solidity_ty, SignatureDataLocation::Memory)); // memory is used by default
+        }
+        let mut ret_type_lst = vec![];
+        for move_ty in fun.get_return_types() {
+            let solidity_ty = self.translate_move_to_solidity_type(&move_ty);
+            ret_type_lst.push((solidity_ty, SignatureDataLocation::Memory));
+        }
+        SoliditySignature {
+            sig_name: fun_name,
+            para_types: para_type_lst,
+            ret_types: ret_type_lst,
+        }
+    }
+
+    /// Parse the solidity signature
+    pub fn parse_into_solidity_signature(
+        &self,
+        sig_str: &str,
+    ) -> anyhow::Result<SoliditySignature> {
+        if let Some(parsed) = SIG_REG.captures(sig_str.trim()) {
+            let sig_name = parsed.name("sig_name").context(PARSE_ERR_MSG)?.as_str();
+            let para_type_str = parsed.name("args").context(PARSE_ERR_MSG)?.as_str();
+            let ret_ty_str_opt = parsed.name("ret_ty");
+            let mut ret_ty = "";
+            if let Some(ret_ty_str) = ret_ty_str_opt {
+                let ret_ty_str_trim = ret_ty_str.as_str().trim();
+                if !ret_ty_str_trim.is_empty() {
+                    let mut parse_error = false;
+                    if let Some(stripped_returns) = ret_ty_str_trim.strip_prefix("returns") {
+                        let stripped_returns_trim = stripped_returns.trim();
+                        if stripped_returns_trim.starts_with('(')
+                            && stripped_returns_trim.ends_with(')')
+                        {
+                            ret_ty = &stripped_returns_trim[1..stripped_returns_trim.len() - 1];
+                        } else {
+                            parse_error = true;
+                        }
+                    } else {
+                        parse_error = true;
+                    }
+                    if parse_error {
+                        return Err(anyhow!(
+                            "error happens when parsing the return types in the signature"
```

Comment:
nit: as you decided to put some error messages into const string, perhaps all of them (or none)?